### PR TITLE
[GH-21] Add support markdown preview in add todo modal

### PR DESCRIPTION
**Pull request by asimsedhain**
#### Summary

This adds support markdown preview in the add todo modal.
temp1
temp2

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-todo/issues/21

### CODE REVIEW

**Review by larkox — APPROVED** [review]
Wonderful work! Thank you so much for the contribution. LGTM.

File: webapp/src/components/root/root.jsx [hunk 1]
```diff
@@ -82,40 +93,90 @@ export default class Root extends React.Component {
...
+                        ) : (
+                            <textarea
+                                className='todoplugin-input'
+                                style={style.textarea}
+                                value={message}
+                                onChange={(e) =>
+                                    this.setState({message: e.target.value})
+                                }
+                            />
+                        )}{' '}
```
Comment (mickmister) [inline webapp/src/components/root/root.jsx:137]: I think this looks better with space on its own line. Also
```suggestion
                        )}
                        {' '}
```

**Review by mickmister — APPROVED** [review]
Some of the multiline formatting changes to the render method's JSX props do not have to do with this PR's functionality. This can mess with the commit tracking of different lines, as it makes it more difficult to quickly access a commit that introduced some functionality.

I also personally think most of the remaining multiline changes they are better to leave single line. It keeps the code more concise to have less lines. The multiline code is straying away from the conventions we usually have in the different javascript projects. Please let me know your thoughts @asimsedhain 🙂

**Review by mickmister — CHANGES_REQUESTED** [review]
Sorry, didn't mean to hit approve. I have one request regarding creating the class names for the buttons, and a few requests regarding formatting.

Thanks for this great improvement @asimsedhain! 🚀

**Comment by asimsedhain** [comment]
Thank you for the feedback, @mickmister.
I forgot I had auto formatted the code. Force of habit.
Also, I have refactored out the class names. Nice catch on that one.
I have incorporated the requested changes. Please let me know your thoughts on it.

File: webapp/src/components/root/root.jsx [hunk 2]
```diff
@@ -85,12 +92,37 @@ export default class Root extends React.Component {
...
-                            value={message}
-                            onChange={(e) => this.setState({message: e.target.value})}
-                        />
+                        <div className='btn-group'>
+                            <button
+                                className={writeButtonClass}
+                                onClick={() => {
+                                    this.setState({previewMarkdown: false});
+                                }}
+                            >{'Write'}</button>
```
Comment (mickmister) [inline webapp/src/components/root/root.jsx:101]:
```suggestion
                            >
                                {'Write'}
                            </button>
```

File: webapp/src/components/root/root.jsx [hunk 3]
```diff
@@ -85,12 +92,37 @@ export default class Root extends React.Component {
...
+                            <button
+                                className={previewButtonClass}
+                                onClick={() => {
+                                    this.setState({previewMarkdown: true});
+                                }}
+                            >{'Preview'}</button>
```
Comment (mickmister) [inline webapp/src/components/root/root.jsx:107]:
```suggestion
                            >
                                {'Preview'}
                            </button>
```

File: webapp/src/components/root/root.jsx [hunk 4]
```diff
@@ -85,12 +92,37 @@ export default class Root extends React.Component {
...
+                        </div>
+                        {this.state.previewMarkdown ? (
+                            <div
+                                className='todoplugin-input'
+                                style={style.markdown}
+                            >{PostUtils.messageHtmlToComponent(
```
Comment (mickmister) [inline webapp/src/components/root/root.jsx:113]:
```suggestion
                            >
                                {PostUtils.messageHtmlToComponent(
```

**Review by mickmister — APPROVED** [review]
Thanks for addressing my comments @asimsedhain! LGTM, I have a few non-blocking suggestions to improve the formatting of embedded text

**Review by aaronrothschild — APPROVED** [review]
Thanks for your contribution @asimsedhain !  Looks great, looking forward to using this - and making it easier for others to format lists :)

**Comment by DHaussermann** [comment]
@asimsedhain This is a very nice enhancement to the pluign. Well done! It seems to work as expected. I find no functional issues here. 
The only follow-up I have is about the CSS styling. It seems like the colors don't adjust to respect the theme the user has applied. This becomes an issue with dark themes.

Screen Shot 2021-02-10 at 3 27 08 PM

I'm not sure if this is an existing issue with the plugin that's more apparent now that you're adding functionality or if there is a way the styling can be made more consistent?

cc @larkox  if you have any thoughts on this.

**Comment by larkox** [comment]
@DHaussermann Looking at the code, it seems that it is indeed taking the colors from the theme. `Mattermost` and `Windows Dark` themes have almost the same button colors, and that may be why the confusion. The "Add Todo" does indeed have a different logic on the theme and should be updated, but it is still going along with the theme. Have you confirmed that with other themes (like `Mattermost Dark`) the color don't change?

**Comment by DHaussermann** [comment]
@larkox yes to clarify, the colours do change when the theme is applied. It's just that ToDo plugin seems to use slightly different colour scheme. I don't think white like on the `Preview` tab above is normally seen on dark theme. Another example is the unselected tab for `Mattermost Dark`
Screen Shot 2021-02-11 at 9 02 21 AM

So, I was curious if this colour selection is defined explicitly in this PR orif this is more of an issue with the ToDo Css in general.

**Comment by DHaussermann** [comment]
Gentle ping on this @larkox 
Is the CSS styling separate form this PR? If so, I'd be happy to create a new issue to address it and we can prceed with this PR now.

**Comment by larkox** [comment]
I think we can assume it is separated from this PR indeed. Thanks @DHaussermann !

**Review by DHaussermann — APPROVED** [review]
Tested and passed
Tested various markdown types and the preview is working as expected.
Added TM4J test case to cover this for release testing

Crated follow-up issue here to address some theme colours 
https://github.com/mattermost/mattermost-plugin-todo/issues/140 


Thanks @asimsedhain this is a great enhancement!

**Pull request merged**